// vuepress-plugin-chart/lib/client/Chart.js
/* eslint-disable no-new */
import Chart from "chart.js/auto";
import { defineComponent, h, onMounted, ref } from "vue";
import JSONfn from "./json-fn";
import "./style.css";
Chart.defaults.maintainAspectRatio = false;
export default defineComponent({
    name: "Chart",
    props: {
        id: {
            type: String,
            required: true
        }
    },
    setup(props) {
        const chartCanvasElement = ref(null);
        const chartElement = ref(null);
        onMounted(() => {
            var _a, _b;
            const content = ((_a = chartElement.value) === null || _a === void 0 ? void 0 : _a.dataset.code) || "";
            const data = JSONfn.parse(content);
            const ctx = (_b = chartCanvasElement.value) === null || _b === void 0 ? void 0 : _b.getContext("2d");
            new Chart(ctx, data);
        });
        return () => h("div", {
            class: "vuepress-chartjs",
            ref: chartElement
        }, [
            h("canvas", {
                ref: chartCanvasElement,
                id: props.id,
                height: 400
            })
        ]);
    }
});
